addons/knobs/src/index.js

```diff
@@ -71,16 +73,22 @@ const defaultOptions = {
   escapeHTML: true,
 };
 
-export const withKnobsOptions = (options = {}) => storyFn => {
-  const allOptions = { ...defaultOptions, ...options };
-
-  manager.setOptions(allOptions);
-  const channel = addons.getChannel();
-  manager.setChannel(channel);
-  channel.emit('addon:knobs:setOptions', allOptions);
-
-  registerKnobs();
-  return storyFn();
-};
-
-export const withKnobs = withKnobsOptions();
+export const withKnobs = makeDecorator({
+  name: 'withKnobs',
+  parameterName: 'knobs',
+  skipIfNoParametersOrOptions: false,
+  wrapper: (getStory, context, { options, parameters }) => {
+    const storyOptions = parameters || options;
+    const allOptions = { ...defaultOptions, ...storyOptions };
+
+    manager.setOptions(allOptions);
+    const channel = addons.getChannel();
+    manager.setChannel(channel);
+    channel.emit('addon:knobs:setOptions', allOptions);
+
+    registerKnobs();
+    return getStory(context);
+  },
+});
+
+export const withKnobsOptions = deprecate(withKnobs, 'Use withKnobs directly');
```

Comment:
Maybe "`withKnobsOptions` is deprecated. Instead, you can pass options into `withKnobs(options)` directly, or use the `knobs` parameter."